space-concordia-robotics/robotics-prototype
# [#284] Implement local mock parameter publisher for battery_voltage, battery_temps and wheel_motor_currents

## Description
The user can set parameters for battery_voltage, battery_temps, and wheel_motor_currents. There are 1 voltage parameter, 3 temperature parameters, and 6 current parameters. Each parameter can be set to rise, remain stable, or fall, and each value published for the parameter suffers noise. The parameters reach their assigned states incrementally based on a user defined rate. 

### Steps for Testing
1. Run `rosgui` and `startgui` in two terminal windows simultaneously.
2. Open the GUI on http://localhost:5000.
3. Change directories to `robotics-prototype/robot/util/rosRoverStart` in another terminal window.
4. Run local mock values using `roslaunch local_rover.launch`.
5. Observe the published values either through the same terminal window as the one launching `roslaunch local_rover.launch` or directly through the GUI. 
6. Observe list of parameters to write to using `rosparam list` . Note: The parameters are called PDS_mock_voltage, PDS_mock_temp# (# is 1 to 3), and PDS_mock_wheel#_current (# is 1 to 6).
7. Set a parameter to rise, constant or fall using `rosparam set /<parameter> <state>`. Note: The parameter states are rise, constant and fall. 
8. Run `rostopic echo <parameter>` in another terminal window to see a specific parameter that is published. 
9. Observe that battery_voltage can be set to 20, 15 and 10, battery_temps to 100, 25, -20, and wheel_motor_currents to 1, 0.3 and 0. 
- **Note**: outputs for all parameters have a difference of +- their noise value. **However**, if the lowest value the variable can reach is 0, therefore the value cannot descend below 0. 
9. Can set different float values for the voltage, temperature and current increment/decrement rates and noise intensities.
10. Set a condition or float value in the `roslaunch local_rover.launch` file for each parameter in order for the simulator to produce the written conditions on startup. 

### **Important Notes**
**1.** If the user removes a parameter from the launch file, the simulation sets the value for that parameter to 0.
- **Example:** `"PDS_mock_voltage"` is never set initialized in launch file, then the GUI will display 0. 

**CRUCIAL NOTE WHEN RUNNING WITH MISSING PARAMETER:** If the user attempts to set the uninitialized parameter from CLI using: `rosparam set /"PDS_mock_voltage" <state>`, the CLI will not display the parameter name from the list of rosparams. 
**For example**, `"PDS_mock_voltage"` is NOT seen as an option from the list of rosparams after. 
**However**, if the user already knows the parameter name by heart regardless, and runs the set command with `"PDS_mock_voltage"`, the next time the launch file is started, then `"PDS_mock_voltage"` will appear as an existing parameter from the rosparam list, and the parameter value on the GUI will be set to the state the user set. This is because the user MANUALLY set the parameter to exist, even though it should not. 

- **Example:** `rosparam set /"PDS_mock_voltage" rise` will set `"PDS_mock_voltage"` to `rise` instead of 0 when the user stops and relaunches the launch file. 
- **Solution:** To reset the parameter to non-existent, the user must stop and restart the gui with `rosgui` and `startgui`


Closes #284
